lib/bamboo/adapters/sendin_blue_adapter_v3.ex

```diff
@@ -0,0 +1,216 @@
+    |> put_cc(email)
+    |> put_bcc(email)
+    |> put_subject(email)
+    |> put_html_body(email)
+    |> put_text_body(email)
+    |> put_attachments(email)
+  end
+
+  defp put_sender(body, %Email{from: {nil, address}}) do
+    body |> Map.put(:sender, %{email: address})
+  end
+
+  defp put_sender(body, %Email{from: {name, address}}) do
+    body |> Map.put(:sender, %{email: address, name: name})
+  end
+
+  defp put_to(body, %Email{to: to}) do
+    body |> put_addresses(:to, address_map(to))
+  end
+
+  defp put_cc(body, %Email{cc: []}), do: body
+
+  defp put_cc(body, %Email{cc: cc}) do
+    body |> put_addresses(:cc, address_map(cc))
+  end
+
+  defp put_bcc(body, %Email{bcc: []}), do: body
+
+  defp put_bcc(body, %Email{bcc: bcc}) do
+    body |> put_addresses(:bcc, address_map(bcc))
+  end
+
+  defp put_subject(body, %Email{subject: subject}), do: Map.put(body, :subject, subject)
+
+  defp put_html_body(body, %Email{html_body: nil}), do: body
+
+  defp put_html_body(body, %Email{html_body: html_body}),
+    do: Map.put(body, :htmlContent, html_body)
+
+  defp put_text_body(body, %Email{text_body: nil}), do: body
+
+  defp put_text_body(body, %Email{text_body: text_body}),
+    do: Map.put(body, :textContent, text_body)
+
+  defp put_reply_to(body, %Email{headers: headers} = _email) do
+    body |> put_reply_to_email(headers) |> put_reply_to_name(headers)
+  end
+
+  defp put_reply_to_email(body, %{"reply-to-email" => email}) do
+    reply_to = body |> Map.get(:replyTo, %{}) |> Map.put(:email, email)
+    Map.put(body, :replyTo, reply_to)
+  end
+
+  defp put_reply_to_email(body, _), do: body
+
+  defp put_reply_to_name(body, %{"reply-to-name" => name}) do
+    reply_to = body |> Map.get(:replyTo, %{}) |> Map.put(:name, name || "")
+    Map.put(body, :replyTo, reply_to)
+  end
+
+  defp put_reply_to_name(body, _), do: body
+
+  defp put_addresses(body, field, addresses), do: Map.put(body, field, addresses)
+
+  defp base_uri do
+    Application.get_env(:bamboo, :sendinblue_base_uri) || default_base_uri()
+  end
+
+  defp put_attachments(body, %Email{attachments: []}), do: body
+
+  defp put_attachments(body, %Email{attachments: atts}) do
+    attachments =
+      atts
+      |> Enum.map(fn attachment -> prepare_attachment(attachment) end)
+
+    Map.put(body, :attachment, attachments)
+  end
+
+  defp prepare_attachment(%Attachment{data: data, filename: filename}) when not is_nil(data) do
+    %{content: Base.encode64(data), name: filename}
+  end
+
+  defp prepare_attachment(%Attachment{path: path, filename: filename} = att)
+       when not is_nil(filename) do
+    case URI.parse(path) do
+      %URI{scheme: nil} ->
+        att |> Map.put(:data, File.read!(path)) |> prepare_attachment()
+
+      %URI{} ->
+        %{url: path, name: filename}
+    end
+  end
+
+  defp prepare_attachment(%Attachment{path: path} = att) when not is_nil(path) do
+    att |> Map.put(:filename, Path.basename(path)) |> prepare_attachment()
```

Comment:
Could you please elaborate on attachments recursive call adding more tests explaining what's going on?

Reply:
Sure, I'll add more test.
In the meantime, these are the ideas behind the recursive calls:

1. If the user does not specify the `filename` parameter, this field will be deduced by the value of `path` (lines 202-203). The recursive call will match one of the following points.
2. If the `data` field is specified, then this information will be interpreted as the binary contents of the attachment (lines 187-189)
3. Otherwise, if the `path` field is specified, the code will try to parse it as a URL. If this fails, then `path` is interpreted as a local path, and the contents of the file will be loaded in the `data` field (lines 194-195, to be matched by the recursive call as specified in 2.). If the URL parsing is successful, then the attachment is handled as a "link" attachment (lines 197-198); the SendInBlue backend will fetch the link and attach the contents to the email.